File: samplestore/src/main/java/com/stripe/samplestore/StoreLineItem.java
package com.stripe.samplestore;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

/**
 * Represents a single line item for purchase in this store.
 */
public class StoreLineItem implements Parcelable {

    @NonNull private final String mDescription;
    private final int mQuantity;
    private final long mUnitPrice;

    StoreLineItem(@NonNull String description, int quantity, long unitPrice) {
        mDescription = description;
        mQuantity = quantity;
        mUnitPrice = unitPrice;
    }

    @NonNull
    String getDescription() {
        return mDescription;
    }

    int getQuantity() {
        return mQuantity;
    }

    long getUnitPrice() {
        return mUnitPrice;
    }

    long getTotalPrice() {
        return mUnitPrice * mQuantity;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeString(this.mDescription);
        dest.writeInt(this.mQuantity);
        dest.writeLong(this.mUnitPrice);
    }

    protected StoreLineItem(@NonNull Parcel in) {
        this.mDescription = in.readString();
        this.mQuantity = in.readInt();
        this.mUnitPrice = in.readLong();
    }

    public static final Parcelable.Creator<StoreLineItem> CREATOR =
            new Parcelable.Creator<StoreLineItem>() {
                @Override
                public StoreLineItem createFromParcel(Parcel source) {
                    return new StoreLineItem(source);
                }

                @Override
                public StoreLineItem[] newArray(int size) {
                    return new StoreLineItem[size];
                }
            };
}

File: samplestore/src/main/java/com/stripe/samplestore/StoreLineItem.kt
package com.stripe.samplestore

import android.os.Parcel
import android.os.Parcelable

/**
 * Represents a single line item for purchase in this store.
 */
class StoreLineItem : Parcelable {

    val description: String
    val quantity: Int
    val unitPrice: Long

    internal val totalPrice: Long
        get() = unitPrice * quantity

    internal constructor(description: String, quantity: Int, unitPrice: Long) {
        this.description = description
        this.quantity = quantity
        this.unitPrice = unitPrice
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(this.description)
        dest.writeInt(this.quantity)
        dest.writeLong(this.unitPrice)
    }

    private constructor(input: Parcel) {
        this.description = input.readString()!!
        this.quantity = input.readInt()
        this.unitPrice = input.readLong()
    }

    companion object CREATOR : Parcelable.Creator<StoreLineItem> {
        override fun createFromParcel(parcel: Parcel): StoreLineItem {
            return StoreLineItem(parcel)
        }

        override fun newArray(size: Int): Array<StoreLineItem?> {
            return arrayOfNulls(size)
        }
    }
}
